doc/private/man/man3/CliManager.3
.TH "CliManager" 3 "Wed Feb 8 2012" "Version 0.0.0" "EMI File Transfer Service (FTS 3)" \" -*- nroff -*-
.ad l
.nh
.SH NAME
CliManager \- 
.SH SYNOPSIS
.br
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "void \fBsetInterface\fP (string interface)"
.br
.ti -1c
.RI "string \fBgetPassword\fP ()"
.br
.ti -1c
.RI "int \fBisChecksumSupported\fP ()"
.br
.ti -1c
.RI "int \fBisDelegationSupported\fP ()"
.br
.ti -1c
.RI "int \fBisRolesOfSupported\fP ()"
.br
.ti -1c
.RI "int \fBisSetTCPBufferSupported\fP ()"
.br
.ti -1c
.RI "int \fBisExtendedChannelListSupported\fP ()"
.br
.ti -1c
.RI "int \fBisChannelMessageSupported\fP ()"
.br
.ti -1c
.RI "int \fBisUserVoRestrictListingSupported\fP ()"
.br
.ti -1c
.RI "int \fBisVersion330StatesSupported\fP ()"
.br
.ti -1c
.RI "int \fBisItVersion330\fP ()"
.br
.ti -1c
.RI "int \fBisItVersion340\fP ()"
.br
.ti -1c
.RI "int \fBisItVersion350\fP ()"
.br
.ti -1c
.RI "int \fBisItVersion370\fP ()"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBCliManager\fP * \fBgetInstance\fP ()"
.br
.in -1c
.SS "Data Fields"

.in +1c
.ti -1c
.RI "string \fBFTS3_PARAM_GRIDFTP\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_MYPROXY\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_DELEGATIONID\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_SPACETOKEN\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_SPACETOKEN_SOURCE\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_COPY_PIN_LIFETIME\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_LAN_CONNECTION\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_FAIL_NEARLINE\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_OVERWRITEFLAG\fP"
.br
.ti -1c
.RI "string \fBFTS3_PARAM_CHECKSUM_METHOD\fP"
.br
.ti -1c
.RI "string \fBFTS3_CA_SD_TYPE\fP"
.br
.ti -1c
.RI "string \fBFTS3_SD_ENV\fP"
.br
.ti -1c
.RI "string \fBFTS3_SD_TYPE\fP"
.br
.ti -1c
.RI "string \fBFTS3_IFC_VERSION\fP"
.br
.ti -1c
.RI "string \fBFTS3_INTERFACE_VERSION\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Definition at line 14 of file CliManager.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for EMI File Transfer Service (FTS 3) from the source code.


fts2-prototype/RELEASE-NOTES
*** Release glite-data-transfer-cli_R_4_0_2 2011-10-14
	#109 	glite-transfer-submit man page issues https://svnweb.cern.ch/trac/glitefts/ticket/109

*** Release glite-data-transfer-cli_R_4_0_0 2011-03-07

    o glite-transfer-channel-audit fails with Segmentation Fault  (bug#68524)
    o Tool to modified automatically the state of the channels given an SE  (bug#61780)    


*** Release glite-data-transfer-cli_R_3_7_5_1 2010-11-29

    o changed util.c to output from stdout to stderr  (bug#75722)

*** Release glite-data-transfer-cli_R_3_7_4_1 2010-01-22

    o added glite-transfer-se-set (bug#61780)

*** Release glite-data-transfer-cli_R_3_7_3_1 2009-12-17

	o Fix for bug #52416: Bad error treatment for ExistsException on the client	side
	o Minimal Python API with test.

*** Release glite-data-transfer-cli_R_3_7_2_1 2009-08-28

    o bug #41584: [delegation] IPv6 bug: non compliant name resolving 
      function in source code (gethostbyname_r)
      This bug also applies to this component. 

*** Release glite-data-transfer-cli_R_3_7_1_1 2009-08-05

	o Fixing a regression: overwrite flag (-o) should not require an argument
    o Updated the test suite to the latest FTS service.

*** Release glite-data-transfer-cli_R_3_7_0_1 2009-07-03

	o v3.7.0 of the interface with checksum support.

	  glite-transfer-submit can take the -K option for checksum comparison
	  and can pass on optional checksums (algorithm:hexadecimal-numbers)
	  for each source-destination SURL pair given by the command line or
	  in a job description file.

*** Release glite-data-transfer-cli_R_3_6_3_1 2009-03-30
	o fixed bug #48684: transfer-cli should obsolete transter-api-c and srm-cli packages

*** Release glite-data-transfer-cli_R_3_6_2_1 2009-03-22
    
	o Fixed Savannah bug #35997:  FTS cli: if an invalid value is passed
	  with -s option, the default is used
    o Removed obsolete commands: glite-transfer-placement-submit
      and glite-transfer-discovery

*** Release glite-data-transfer-cli_R_3_6_0_1 2009-01-09
    o fixed bug 45167
    o fixed bug 28212
    o fixed bug #39819
	o added site group management tools
	o Fixed the seg fault for --tx-log-level and --srmcopy-direction options
	o Fixed the option --urlcopy-tx-to-per-mb that was executing --urlcopy-no-prog-to
	o Fixed libraries build order (avoid relinking and appending U to .so file names)

*** Release glite-data-transfer-cli_R_3_5_2_1 2008-12-04
	o Fixed bug 41682

*** Release glite-data-transfer-cli_R_3_5_1_1 2008-12-03
	o slc5 fixes

*** Release glite-data-transfer-cli_R_3_5_0_1 2008-11-27
    o Merged transfer-api-c into transfer-cli.
    o updated to schema version 3.5.0

*** Release glite-data-transfer-cli_R_3_4_5_1 2008-08-11
    o fixed bug #40064 
	o fixed bug #39538
    o refactored options parsing 
    o modified help messages

*** Release glite-data-transfer-cli_R_3_4_4_3 2008-06-17
    o Added dependency from gsoap

*** Release glite-data-transfer-cli_R_3_4_4_2 2008-05-08
	o Fix bug #33406

*** Release glite-data-transfer-cli_R_3_4_3_2 2008-03-20
	o changes to command line help in submit - clearer token handling

*** Release glite-data-transfer-cli_R_3_4_3_1 2008-03-18
	 #30034: FTS: cannot set numerical channel properties to zero
	 #33406: FTS: CLI version string is not reported correctly in verbose option
	 #28524: Missing -o parameter from glite-transfer-submit help
	 Interface changes related to the latest schema changes 3.2.0
	 Number of undocumented bugs fixed related to the marshalling.

*** Release glite-data-transfer-cli_R_3_4_2_2 2007-06-06
    o lib64 changes

*** Release glite-data-transfer-cli_R_3_4_2 2007-06-01

   o Added overwrite flag for SRM 2.2. Proper name for space token description.

*** Release glite-data-transfer-cli_R_3_4_1-1 2007-04-20

   o Fixed SLC4 build error and a few compiler warnings.

*** Release glite-data-transfer-cli_R_3_3_0-4 2007-03-02
   o Compile with gsoap 2.7.6b and 2.7.9b 

*** Release glite-data-transfer-cli_R_3_3_0-1 2006-08-22
   o Add option -x to the glite-transfer-discovery tool in order to display all the 
     FTS service information
   o Add option --all to the glite-transfer-discovery tool in order to return
     all the FTS services that are able to manage a transfer between the given two sites
   o Fix for #19196: FTS: cli should support site names in teh -s option

*** Release glite-data-transfer-cli_R_3_2_3_4 2006-07-06

   o Fix for #17975 - service info interface version bad
   o Fix for #16546 - bad error on addChannel
   o Make interesting states (default serach) only Active, Submitted and Pending
   o Better date format on stats
   o Fix for #16581 - unclear manpage for addManager

*** Release glite-data-transfer-cli_R_3_2_3-3 2006-06-21

	o Merge changes from 3.2.2-3

*** Release glite-data-transfer-cli_R_3_2_2-3 2006-06-21
    Fixed Bugs:
     o #17608: FTS: glite-transfer-stats-vo segfaults

*** Release glite-data-transfer-cli_R_3_2_3-2 2006-06-14

	o Merge changes from 3.2.2-2

*** Release glite-data-transfer-cli_R_3_2_2-2 2006-06-14
    Fixed Bugs:
     o #17460: FTS: glite-transfer-discovery display the wrong destination site when the --verbose option is used

*** Release glite-data-transfer-cli_R_3_2_3 2006-05-02

	o Added Doxygen API documentation for the simple transfer API.

*** Release glite-data-transfer-cli_R_3_2_2 2006-04-28
    Add command line options -T and -S to glite-transfer-discovery
    Fixed Bugs:
     o #16424 FTS: BDII service discovery should work 
     o #16215: fix man page of glite-transfer-channel-list 

*** Release glite-data-transfer-cli_R_3_2_1 2006-03-23

    o CGSI_gSOAP library location has changed from ${repository}/${ext.cgsi-gsoap.subdir} to ${with.cgsi-gsoap.prefix}

*** Release glite-data-transfer-cli_R_3_2_0 2006-02-22
    Add glite-transfer-discovery tool for returning the endpoint of an
       FTS service that manage a channel between two sites
    Fixed bugs:
     o #14765 FTS: Logical name is not displayed for placement jobs


src/cli/CliManager.cpp
/*
 * CliManager.cpp
 *
 *  Created on: Feb 7, 2012
 *      Author: simonm
 */

#include "CliManager.h"

#include <termios.h>
#include <iostream>
#include <boost/tokenizer.hpp>
#include <boost/lexical_cast.hpp>
using namespace boost;

CliManager::CliManager() {

	FTS3_PARAM_GRIDFTP = "gridftp";
	FTS3_PARAM_MYPROXY = "myproxy";
	FTS3_PARAM_DELEGATIONID = "delegationid";
	FTS3_PARAM_SPACETOKEN = "spacetoken";
	FTS3_PARAM_SPACETOKEN_SOURCE = "source_spacetoken";
	FTS3_PARAM_COPY_PIN_LIFETIME = "copy_pin_lifetime";
	FTS3_PARAM_LAN_CONNECTION = "lan_connection";
	FTS3_PARAM_FAIL_NEARLINE = "fail_nearline";
	FTS3_PARAM_OVERWRITEFLAG = "overwrite";
	FTS3_PARAM_CHECKSUM_METHOD = "checksum_method";

	FTS3_CA_SD_TYPE = "org.glite.ChannelAgent";
	FTS3_SD_ENV = "GLITE_SD_FTS_TYPE";
	FTS3_SD_TYPE = "org.glite.FileTransfer";
	FTS3_IFC_VERSION = "GLITE_FTS_IFC_VERSION";
	FTS3_INTERFACE_VERSION = "3.7.0"; // TODO where it comes from?
}

CliManager::CliManager(CliManager const&) {

}

CliManager& CliManager::operator=(CliManager const&) {
	return *this;
}

CliManager::~CliManager() {
	delete manager;
}

CliManager* CliManager::manager = 0;

CliManager* CliManager::getInstance() {

	if (!manager) {
		manager = new CliManager();
	}
	return CliManager::manager;
}

void CliManager::setInterface(string interface) {

	if (interface.empty()) return;


	char_separator<char> sep(".");
	tokenizer< char_separator<char> > tokens(interface, sep);
	tokenizer< char_separator<char> >::iterator it = tokens.begin();

	if (it == tokens.end()) return;

	string s = *it++;
	major = lexical_cast<long>(s);

	if (it == tokens.end()) return;

	s = *it++;
	minor = lexical_cast<long>(s);

	if (it == tokens.end()) return;

	s = *it;
	patch = lexical_cast<long>(s);
}

string CliManager::getPassword() {

    termios stdt;
    // get standard cmd settings
    tcgetattr(STDIN_FILENO, &stdt);
    termios newt = stdt;
    // turn off echo while typing
    newt.c_lflag &= ~ECHO;
    if (tcsetattr(STDIN_FILENO, TCSANOW, &newt)) {
    	cout << "submit: could not set terminal attributes" << endl;
    	tcsetattr(STDIN_FILENO, TCSANOW, &stdt);
    	return "";
    }

    string pass1, pass2;

    cout << "Enter MyProxy password: ";
    cin >> pass1;
    cout << endl << "Enter MyProxy password again: ";
    cin >> pass2;
    cout << endl;

    tcsetattr(STDIN_FILENO, TCSANOW, &stdt);

    if (pass1.compare(pass2)) {
    	cout << "Entered MyProxy passwords do not match." << endl;
    	return "";
    }

    return pass1;
}

int CliManager::isChecksumSupported() {
    return isItVersion370();
}

int CliManager::isDelegationSupported() {
    return isItVersion330();
}

int CliManager::isRolesOfSupported() {
    return isItVersion330();
}

int CliManager::isSetTCPBufferSupported() {
    return isItVersion330();
}

int CliManager::isExtendedChannelListSupported() {
    return isItVersion330();
}

int CliManager::isChannelMessageSupported() {
    return isItVersion330();
}

int CliManager::isUserVoRestrictListingSupported() {
    return isItVersion330();
}

int CliManager::isVersion330StatesSupported(){
    return isItVersion330();
}

// TODO !!!

int CliManager::isItVersion330() {
    return major >= 3 && minor >= 3 && patch >= 0;
}

int CliManager::isItVersion340() {
    return major >= 3 && minor >= 4 && patch >= 0;
}

int CliManager::isItVersion350() {
    return major >= 3 && minor >= 5 && patch >= 0;
}

int CliManager::isItVersion370() {
	return major >= 3 && minor >= 7 && patch >= 0;
}

